src/datahike/api.cljc

```diff
@@ -21,948 +25,12 @@
-
-             Useful patterns:
-
-                 ; find all entities with age in a specific range (inclusive)
-                 (->> (index-range db {:attrid :age :start 18 :end 60}) (map :e))"}
-  index-range
-  (fn [db {:keys [attrid start end]}]
-    (dbi/-index-range db attrid start end)))
-
-(s/fdef
-  listen
-  :args (s/alt :no-key (s/cat :conn spec/SConnection :callback fn?)
-               :with-key (s/cat :conn spec/SConnection :key any? :callback fn?))
-  :ret any?
-  :fn #(if (= :with-key (-> % :args first))
-         (= (:ret %) (-> % :args second :key))
-         true))
-(def ^{:arglists '([conn callback] [conn key callback])
-       :doc "Listen for changes on the given connection. Whenever a transaction is applied to the database via
-             [[transact]], the callback is called with the transaction report. `key` is any opaque unique value.
-
-             Idempotent. Calling [[listen]] with the same twice will override old callback with the new value.
-
-             Returns the key under which this listener is registered. See also [[unlisten]]."}
-  listen
-  dcore/listen!)
-
-(s/fdef
-  unlisten
-  :args (s/cat :conn spec/SConnection :key any?)
-  :ret map?)
-(def ^{:arglists '([conn key])
-       :doc "Removes registered listener from connection. See also [[listen]]."}
-  unlisten
-  dcore/unlisten!)
-
-(s/fdef
-  schema
-  :args (s/cat :db spec/SDB)
-  :ret spec/SSchema)
-(defn ^{:arglists '([db])
-        :doc "Returns current schema definition."}
-  schema
-  [db]
-  (reduce-kv
-   (fn [m k v]
-     (cond
-       (and (keyword? k)
-            (not (or (ds/entity-spec-attr? k)
-                     (ds/schema-attr? k)
-                     (ds/sys-ident? k)))) (update m k #(merge % v))
-       (number? k) (update m v #(merge % {:db/id k}))
-       :else m))
-   {}
-   (dbi/-schema db)))
-
-(s/fdef
-  reverse-schema
-  :args (s/cat :db spec/SDB)
-  :ret map?)
-(defn ^{:arglists '([db])
-        :doc "Returns current reverse schema definition."}
-  reverse-schema
-  [db]
-  (reduce-kv
-   (fn [m k v]
-     (let [attrs (->> v
-                      (remove #(or (ds/entity-spec-attr? %)
-                                   (ds/sys-ident? %)
-                                   (ds/schema-attr? %)))
-                      (into #{}))]
-       (if (empty? attrs)
-         m
-         (assoc m k attrs))))
-   {}
-   (dbi/-rschema db)))
-
-(s/fdef
-  metrics
-  :args (s/cat :db spec/SDB)
-  :ret spec/SMetrics)
-(defn ^{:arglists '([db])
-        :doc "Returns database metrics"}
-  metrics
-  [db]
-  (db/metrics db))
-
-(defn ^{:arglists '([])
-        :doc "Loads default config for the current environment"}
-  load-config
-  []
-  (config/load-config))
+(doseq [[n {:keys [args ret fn doc impl]}] api-specification]
+  (eval
+   `(s/fdef ~n :args ~args :ret ~ret ~@(when fn [:fn fn])))
+  (eval
+   `(~'def
+     ~(with-meta n
+        {:arglists `(spec-args->argslist (quote ~args))
+         :doc      doc})
+     ~impl)))
```

Comment:
:smile:  I hope this is also possible for the bindings of the Babashka pod and the cli. The GraalVM is able to deal with these `eval` calls because they happen also at namespace loading time. One downside is that clojure docs will probably not find any of the API functions. An alternative is to manually expand this and paste it in the REPL into the file (or write it to .clj file).

Reply:
you mean cljdoc will probably not work anymore?

Reply:
Yes.